app/upgrades/v9.2.0/upgrades.go

```diff
@@ -15,3 +19,89 @@ func CreateUpgradeHandler(args upgrades.UpgradeHandlerArgs) upgradetypes.Upgrade
 		return args.ModuleManager.RunMigrations(ctx, args.Configurator, vm)
 	}
 }
+
+func Fork(ctx sdk.Context, keepers *stakingkeeper.Keeper) {
+
+	params := keepers.GetParams(ctx)
+	maxValidators := params.MaxValidators
+
+	alldels := keepers.GetAllDelegations(ctx)
+
+	valShareMap := make(map[string]stakingtypes.Validator)
+	for _, del := range alldels {
+		validator, ok := valShareMap[del.ValidatorAddress]
+		if !ok {
+			validator, found := keepers.GetValidator(ctx, del.GetValidatorAddr())
+			if !found {
+				panic("Validator not found" + del.ValidatorAddress)
+			}
+			validator.Tokens = sdk.ZeroInt()
+			validator.DelegatorShares = sdk.ZeroDec()
+			validator.ValidatorBondShares = sdk.ZeroDec()
+			//validator.LiquidShares = sdk.ZeroDec() //Will be refreshed
+			valShareMap[del.ValidatorAddress] = validator
+		}
+		validator.DelegatorShares = validator.DelegatorShares.Add(del.Shares)
+		if del.ValidatorBond {
+			validator.ValidatorBondShares = validator.ValidatorBondShares.Add(del.Shares)
+		}
+		valShareMap[del.ValidatorAddress] = validator
+		// Tokens we do directly in the next.
+	}
+	err := keepers.RefreshTotalLiquidStaked(ctx)
+	if err != nil {
+		panic(err)
+	}
+	allvals := keepers.GetAllValidators(ctx)
+	for _, val := range allvals {
+		calculatedVal, ok := valShareMap[val.OperatorAddress]
+		if !ok {
+			panic("validator not found" + val.OperatorAddress)
+		}
+		if !val.DelegatorShares.Equal(calculatedVal.DelegatorShares) {
+			// SHOW ME
+			ctx.Logger().Info(fmt.Sprintf("Validator %s is affected", val.OperatorAddress))
+			tokens := val.TokensFromShares(calculatedVal.DelegatorShares)
+			calculatedVal.Tokens = tokens.TruncateInt()
+			valShareMap[calculatedVal.OperatorAddress] = calculatedVal
+
+			keepers.SetValidator(ctx, calculatedVal)
+			// Fix voting power
+			fixPower(ctx, keepers, val, calculatedVal, maxValidators)
+		} else {
+			ctx.Logger().Info(fmt.Sprintf("Validator %s is ok", val.OperatorAddress))
+		}
+	}
+
+}
+
+func fixPower(ctx sdk.Context, k *stakingkeeper.Keeper, oldval, newval stakingtypes.Validator, maxValidators uint32) {
+
+	iterator := k.ValidatorsPowerStoreIterator(ctx)
+
+	keys := [][]byte{}
+	values := []sdk.ValAddress{}
+	for count := 0; iterator.Valid() && count < int(maxValidators); iterator.Next() {
+		// everything that is iterated in this loop is becoming or already a
+		// part of the bonded validator set
+		valAddr := sdk.ValAddress(iterator.Value())
+		if newval.GetOperator().Equals(valAddr) {
+			keys = append(keys, iterator.Key())
```

Comment:
make sure to explicitly copy key bytes - KV iterators might have nasty optimizations that reuse key bytes and nuke it after iterator is closed.

Reply:
done! better safe than sorry.